Xeno strains refactor

SabreML opened the pull request:
# About the pull request

Refactors xeno strains by taking out the old `/datum/xeno_mutator`/`/datum/mutator_set` system and replacing it with a new `datum/xeno_strain`, since as far as I'm aware mutators haven't been touched in almost 5 years.

As a side note, the DMI changes are from me deleting the icon states of the unused 'Tremor' Burrower strain, since I figured it fit within the bounds of the PR. (The strain's sprites are identical to that of the base caste.)

# Explain why it's good for the game

The new datum should be a lot more streamlined than the old system, and require much less boilerplate when making a new strain. Ideally this could even replace `/datum/behavior_delagate` with signal handlers in the future, although a few more refactors would be required before that's possible.


# Testing Photographs and Procedure
<details>
<summary>Screenshots & Videos</summary>

Put screenshots and videos here with an empty line between the screenshots and the `<details>` tags.

</details>


# Changelog
:cl:
refactor: Refactored xeno strains.
add: Added a keybind setting to open the 'Purchase Strain' window. (Also renamed the macro)
/:cl:

SabreML commented:
The UI label can be removed now. It was just from me putting this in a jsx file: bd43b18df9c0926706adb3b3740f615ee58e1e38.

Birdtalon commented:
I will try to get through this soon.

cm13-github commented:
This pull request has conflicts, please resolve those before we can evaluate the pull request.

cm13-github commented:
Conflicts have been resolved. A maintainer will review the pull request shortly.

nauticall commented:
please add ingame screenshots of your sprites to the PR description, thanks 🙏

cm13-github commented:
This pull request has conflicts, please resolve those before we can evaluate the pull request.

SabreML commented:
> @nauticall just FYI this PR only deletes unused Tremor burrower sprites.

Yep! I've just added something about it to the description since it isn't obvious from the icon diffs.

cm13-github commented:
Conflicts have been resolved. A maintainer will review the pull request shortly.

Birdtalon commented on code/modules/mob/living/carbon/xenomorph/strains/xeno_strain.dm:
```diff
@@ -0,0 +1,131 @@
+ * Returns a bool indicating if the strain was successfully applied.
+ */
+/datum/xeno_strain/proc/_add_to_xeno(mob/living/carbon/xenomorph/xeno)
+	// Override `apply_changes()`, not this! (Unless you know what you're doing.)
```
This would be useful to have in the code doc above.

Birdtalon commented on code/modules/mob/living/carbon/xenomorph/strains/castes/drone/gardener.dm:
```diff
@@ -368,9 +364,6 @@
 
 	var/mutable_appearance/fruit_sac_overlay_icon
 
-/datum/behavior_delegate/drone_gardener/add_to_xeno()
-	on_update_icons()
-
```
Does this not need to be called somewhere?

SabreML replied:
It's done in `/datum/xeno_strain/proc/_add_to_xeno()` just after the behavior delegate is added:
https://github.com/cmss13-devs/cmss13/blob/e129cfd6494555b5c8b44d529cab7c5258a454d4/code/modules/mob/living/carbon/xenomorph/strains/xeno_strain.dm#L42-L46
`/mob/living/carbon/xenomorph/update_icons()` calls `on_update_icons()` on the behaviour delegate, so it has the same end effect.

(Probably should have mentioned this in the commit or something when I removed `/datum/behavior_delegate/drone_gardener/add_to_xeno()`, since it is a bit obscured)

Birdtalon commented on code/modules/mob/living/carbon/xenomorph/castes/Queen.dm:
```diff
@@ -936,7 +934,7 @@
 	if(Queen.ovipositor)
 		Queen.icon = Queen.queen_ovipositor_icon
-		Queen.icon_state = "[Queen.mutation_icon_state || Queen.mutation_type] Queen Ovipositor"
+		Queen.icon_state = "Normal Queen Ovipositor"
```
Should we use the get strain here in case we ever have queen strains in future?

SabreML replied:
Oh yeah good idea.

Birdtalon commented on code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm:
```diff
@@ -879,13 +862,12 @@
-			var/datum/behavior_delegate/praetorian_warden/BD = X.behavior_delegate
-			if (!istype(BD))
-				return
+		var/datum/behavior_delegate/praetorian_warden/BD = X.behavior_delegate
+		if (!istype(BD))
+			return
```
```suggestion
		var/datum/behavior_delegate/praetorian_warden/behavior = X.behavior_delegate
		if (!istype(behavior))
			return
```

Birdtalon commented on code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm:
```diff
@@ -879,13 +862,12 @@
 
-			if (!BD.use_internal_hp_ability(debuff_cost))
-				return
+		if (!BD.use_internal_hp_ability(debuff_cost))
+			return
```
```suggestion
		if (!behavior.use_internal_hp_ability(debuff_cost))
			return
```

Birdtalon commented on code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm:
```diff
@@ -841,36 +828,32 @@
 			return
 
 		var/bonus_heal = 0
+		var/datum/behavior_delegate/praetorian_warden/BD = X.behavior_delegate
+		if (!istype(BD))
+			return
```
```suggestion
		var/datum/behavior_delegate/praetorian_warden/behavior= X.behavior_delegate
		if (!istype(behavior))
			return
```

Birdtalon commented on code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm:
```diff
@@ -841,36 +828,32 @@
-
-			if (!BD.use_internal_hp_ability(heal_cost))
-				return
+		if (!BD.use_internal_hp_ability(heal_cost))
```
```suggestion
		if (!behavior.use_internal_hp_ability(heal_cost))
```

Birdtalon commented on code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm:
```diff
@@ -841,36 +828,32 @@
-			bonus_heal = BD.internal_hitpoints*0.5
-			if (!BD.use_internal_hp_ability(bonus_heal))
-				bonus_heal = 0
+		bonus_heal = BD.internal_hitpoints*0.5
+		if (!BD.use_internal_hp_ability(bonus_heal))
+			bonus_heal = 0
```
```suggestion
		bonus_heal = behavior.internal_hitpoints*0.5
		if (!behavior.use_internal_hp_ability(bonus_heal))
			bonus_heal = 0
```

Birdtalon commented on code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm:
```diff
@@ -841,36 +828,32 @@
-		if(X.mutation_type == PRAETORIAN_WARDEN)
-			var/datum/behavior_delegate/praetorian_warden/warden_delegate = X.behavior_delegate
-			warden_delegate.transferred_healing += heal_amount
+		BD.transferred_healing += heal_amount
```
```suggestion
		behavior.transferred_healing += heal_amount
```

Birdtalon commented on code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm:
```diff
@@ -806,17 +794,16 @@
-			var/datum/behavior_delegate/praetorian_warden/BD = X.behavior_delegate
-			if (!istype(BD))
-				return
+		var/datum/behavior_delegate/praetorian_warden/BD = X.behavior_delegate
+		if (!istype(BD))
+			return
```
```suggestion
		var/datum/behavior_delegate/praetorian_warden/behavior= X.behavior_delegate
		if (!istype(behavior))
			return
```

Birdtalon commented on code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm:
```diff
@@ -806,17 +794,16 @@
 
-			if (!BD.use_internal_hp_ability(shield_cost))
-				return
+		if (!BD.use_internal_hp_ability(shield_cost))
+			return
```
```suggestion
		if (!behavior.use_internal_hp_ability(shield_cost))
			return
```

Birdtalon commented on code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm:
```diff
@@ -806,17 +794,16 @@
-			bonus_shield = BD.internal_hitpoints*0.5
-			if (!BD.use_internal_hp_ability(bonus_shield))
-				bonus_shield = 0
+		bonus_shield = BD.internal_hitpoints*0.5
+		if (!BD.use_internal_hp_ability(bonus_shield))
+			bonus_shield = 0
```
```suggestion
		bonus_shield = behavior.internal_hitpoints*0.5
		if (!behavior.use_internal_hp_ability(bonus_shield))
			bonus_shield = 0
```

Birdtalon commented on code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm:
```diff
@@ -135,10 +134,9 @@
-			var/datum/behavior_delegate/praetorian_vanguard/BD = X.behavior_delegate
-			if (istype(BD))
-				BD.regen_shield()
+		var/datum/behavior_delegate/praetorian_vanguard/BD = X.behavior_delegate
+		if (istype(BD))
+			BD.regen_shield()
```
```suggestion
		var/datum/behavior_delegate/praetorian_vanguard/behavior= X.behavior_delegate
		if (istype(behavior))
			behavior.regen_shield()
```

Birdtalon commented on code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm:
```diff
@@ -70,10 +70,9 @@
-			var/datum/behavior_delegate/praetorian_vanguard/BD = source_xeno.behavior_delegate
-			if (istype(BD))
-				BD.regen_shield()
+		var/datum/behavior_delegate/praetorian_vanguard/BD = source_xeno.behavior_delegate
+		if (istype(BD))
+			BD.regen_shield()
```
```suggestion
		var/datum/behavior_delegate/praetorian_vanguard/behavior= source_xeno.behavior_delegate
		if (istype(behavior))
			behavior.regen_shield()
```

Birdtalon commented on code/modules/mob/living/carbon/xenomorph/abilities/lurker/lurker_powers.dm:
```diff
@@ -1,3 +1,35 @@
+
+	if(found)
+		var/datum/action/xeno_action/onclick/lurker_invisibility/lurker_invis = get_xeno_action_by_type(xeno, /datum/action/xeno_action/onclick/lurker_invisibility)
+		if(istype(lurker_invis))
+			lurker_invis.invisibility_off()
```
Don't need this check `get_xeno_action_by_type()` already does it.

Birdtalon commented on code/modules/mob/living/carbon/xenomorph/abilities/defender/defender_powers.dm:
```diff
@@ -207,49 +202,53 @@
+		X.small_explosives_stun = TRUE
+
+// Steel crest override
+/datum/action/xeno_action/activable/fortify/steel_crest/apply_modifiers(mob/living/carbon/xenomorph/X, fortify_state)
+	if(fortify_state)
+		X.armor_deflection_buff += 10
+		X.armor_explosive_buff += 60
+		X.ability_speed_modifier += 3
+		X.damage_modifier -= XENO_DAMAGE_MOD_SMALL
+	else
+		X.armor_deflection_buff -= 10
+		X.armor_explosive_buff -= 60
+		X.ability_speed_modifier -= 3
+		X.damage_modifier += XENO_DAMAGE_MOD_SMALL
+
```
`X` to `xeno` please.

Birdtalon commented on code/modules/mob/living/carbon/xenomorph/abilities/defender/defender_powers.dm:
```diff
@@ -207,49 +202,53 @@
+	apply_modifiers(X, fortify_state)
+	X.update_icons()
+
+/datum/action/xeno_action/activable/fortify/proc/apply_modifiers(mob/living/carbon/xenomorph/X, fortify_state)
+	if(fortify_state)
+		X.armor_deflection_buff += 30
+		X.armor_explosive_buff += 60
+		ADD_TRAIT(X, TRAIT_IMMOBILIZED, TRAIT_SOURCE_ABILITY("Fortify"))
+		X.anchored = TRUE
+		X.small_explosives_stun = FALSE
+	else
+		X.armor_deflection_buff -= 30
+		X.armor_explosive_buff -= 60
+		X.small_explosives_stun = TRUE
```
`X` to `xeno` please

Birdtalon commented on code/datums/xeno_shields/shield_types/vanguard_shield.dm:
```diff
@@ -50,7 +50,6 @@
-		var/datum/behavior_delegate/praetorian_vanguard/BD = linked_xeno.behavior_delegate
-		if (istype(BD))
-			BD.last_combat_time = world.time
+	var/datum/behavior_delegate/praetorian_vanguard/BD = linked_xeno.behavior_delegate
+	if (istype(BD))
+		BD.last_combat_time = world.time
```
```suggestion
	var/datum/behavior_delegate/praetorian_vanguard/behavior = linked_xeno.behavior_delegate
	if (istype(behavior))
		behavior.last_combat_time = world.time
```

SabreML commented:
> @SabreML this PR breaks/removes the purchase strains hotkey currently

Shoot, thanks for letting me know!

SabreML commented:
@Megaddd It looks like the issue was from me changing the verb's `name` from 'Purchase Strains' to 'Purchase Strain' (since it's a bit more accurate), which means that the command to activate it also changed to `Purchase-Strain`.

I've just added an "official" keybind setting for it, so I'd suggest either swapping over to that or editing your macro to use the non-plural command.

Birdtalon reviewed (approved):
Looking good, testing issues have been resolved.
